File: socket/tcpProxy.py

svcAddr = '127.0.0.1' ## http://127.0.0.01:8080
svcPort = 8080 ## netstat -a -n -o | find "8080"
upstrmServer = "192.168.43.1"
upstrmPort = 80
httpContent = True

import socket
import datetime

def printHeadTail(data, excerptLength) :
  tailEnd = len(data)
  tailStart = (tailEnd-excerptLength) if (tailEnd>=excerptLength) else 0
  if tailStart>0 : print(repr(data[0:tailEnd-tailStart]))
  print(repr(data[tailStart:tailEnd]))
#

if __name__ == '__main__':
  tcpSockTyp = socket.SOCK_STREAM ## udp: socket.SOCK_DGRAM | tcp:socket.SOCK_STREAM
  if True :
  #try:
    svcSock = socket.socket(socket.AF_INET, tcpSockTyp)
    print("starting "+svcAddr+":"+str(svcPort)+" ... "+str(svcSock.type))
    svcSock.bind((svcAddr, svcPort))  ## netstat -ano | findstr 53
    svcSock.listen(2)
    while True:
      clientSock, clientAddr = svcSock.accept(); 
      clientSock.settimeout(1)
      dataAll = b""
      try:
        chnkSz = 4096
        while True:
          data = clientSock.recv(chnkSz)
          if (not data) or len(data)==0 : break
          dataAll = dataAll + data
          #if (len(data)<chnkSz) : break
          if httpContent and data.endswith(b"\r\n\r\n") : break
        #
      except Exception as e:
        print(e)
      #
      print(
        "\n" + datetime.datetime.now().strftime("%H:%M:%S")
        + " [svcSock:recv] " + str(clientAddr)
        + " (" + str(len(dataAll)) + " bytes)..."
      )
      printHeadTail(dataAll, 50)
      #print(repr(dataAll))

      with socket.socket(socket.AF_INET, tcpSockTyp) as upstrmSock :
        upstrmSock.connect((upstrmServer, upstrmPort))
        upstrmSock.settimeout(5)
        upstrmSock.sendall(dataAll)
        upstrmRspAll = b""
        try:
          chnkSz = 4096
          headerEnd = -1
          contentLength = -1
          while True:
            upstrmRsp = upstrmSock.recv(chnkSz)
            if (not upstrmRsp) or len(upstrmRsp)==0 : break
            upstrmRspAll = upstrmRspAll + upstrmRsp

            if httpContent and (headerEnd < 0) :
              headerEnd = upstrmRspAll.find(b"\r\n\r\n")
              if (headerEnd > 0) :
                headerEnd = headerEnd + 4
                print(" **** headerEnd : "+str(headerEnd))
                header = upstrmRspAll[0:headerEnd].decode("utf8")
                contentLength = header.lower().find("content-length:")
                if (contentLength > 0) :
                  contentLength = header[contentLength+15:header.find("\r\n",contentLength)]
                  print(" *** contentLength: "+str(contentLength))
                  contentLength = int(contentLength)
                #
              #
            #
            #if (len(upstrmRsp)<chnkSz) : break
            if httpContent :
              if upstrmRsp.endswith(b"\r\n\r\n") : break
              if contentLength>0 and len(upstrmRspAll)>=(headerEnd+contentLength) : break
            #
          #
        except Exception as e:
          print(e)
        #
        print(
          datetime.datetime.now().strftime("%H:%M:%S")
          + " [upstrmSock:recv] " + str(upstrmServer) + ":" + str(upstrmPort)
          + " (" + str(len(upstrmRspAll)) + " bytes)..."
        )
        printHeadTail(upstrmRspAll, 50)
        #print(repr(upstrmRspAll))

        ##upstrmRsp = b"HTTP/1.1 200 OK\r\nContent-Type: text/plain\n\nOK"
        clientSock.sendall(upstrmRspAll);
      #
      clientSock.close()
    #
  #except Exception as e:
  #  print(e)
  #  svcSock.close()
  #
#

File: socket/tcpProxy.lua
#!/usr/bin/env lua

local svcAddr = '*' -- '127.0.0.1' -- http://127.0.0.01:8080
local svcPort = 8080 -- netstat -a -n -o | find "8080"
local upstrmServer = "192.168.43.1"
local upstrmPort = 80
local httpContent = true

------------------------------------------------------------------------

local function str2Hex(s)
  local bytes = {}
  for i=1,s:len() do
    bytes[#bytes+1] = ('%02x'):format(s:byte(i,i))
  end
  return table.concat(bytes, '')
end

local function printHeadTail(data, excerptLength, showHex)
  local tailEnd = #data
  local tailStart = (tailEnd-excerptLength)
  if tailStart<1 then tailStart=1 end
  local tailData = data:sub(tailStart,tailEnd)
  if tailStart>1 then
    --tailData = data:sub(1,tailEnd-tailStart) .. string.rep(string.char(0),4) .. tailData
    tailData = data:sub(1,tailEnd-tailStart) .. "<<<>>>" .. tailData
  end
  if showHex then tailData=str2Hex(tailData) end
  print(tailData)
end

local function scktReceiveData(sckt, chnkSz, nTimeout, dataType)
  local count = 0
  local dataAll = ""
  repeat
    count = count + 1
    local rcvData,rcvErr,rcvPartial = sckt:receive(chnkSz)
    -- default :receive('*l'): read text line !!
    -- :receive('*a') not working here ??!
    sckt:settimeout(nTimeout) -- shorten timeout for subsequent data
    dataAll = dataAll .. (rcvData or rcvPartial)
    if (rcvErr~=nil) then
      print("count:"..count.." rcvErr:"..rcvErr)
      break
    --elseif (rcvData:sub(-1):byte()==10) then
    --  print("count:"..count.." <eof>"..)
    --  break
    end
  until (rcvErr~=nil)
  return dataAll
end

------------------------------------------------------------------------

local socket = require("socket")

local svcSckt = assert(socket.bind(svcAddr, svcPort))
local svcAddr1,svcPort1 = svcSckt:getsockname()
svcSckt:settimeout(10)

print("Beginning server loop ["..svcAddr1..":"..svcPort1.."] ...")
while true do
  local clientSckt,err = svcSckt:accept()
  if clientSckt then
    local time = os.date("*t")
    print(("\n%02d:%02d:%02d [svcSckt:accept] %s"):format(
      time.hour, time.min, time.sec, clientSckt:getpeername()
    ))
    clientSckt:settimeout(3)
    local data =  scktReceiveData(clientSckt, 64, 1, "HTTP")
    if data then
      local clientAddr,clientPort = clientSckt:getsockname()
      print(("[clientSckt:receive] (%d bytes)"):format(
        string.len(data)
      ))
      --print(str2Hex(data));  --print(data)
      --printHeadTail(data, 50, true)
      printHeadTail(data, 35, false)

      -- data = "GET / HTTP/1.1\r\n\r\n"

      local upstrmSckt = assert(socket.tcp())
      upstrmSckt:settimeout(5)
      upstrmSckt:connect(upstrmServer, upstrmPort)
      upstrmSckt:send(data.."\r\n\r\n")
      local upstrmRsp = scktReceiveData(upstrmSckt, 128, 1, "HTTP")
      upstrmSckt:close()

      local time = os.date("*t")
      print(("%02d:%02d:%02d [upstrmSckt:receivefrom] %s:%s (%d bytes)"):format(
        time.hour, time.min, time.sec, upstrmServer, upstrmPort, string.len(upstrmRsp)
      ))
      --print(str2Hex(upstrmRsp));  --print(upstrmRsp)
      --printHeadTail(upstrmRsp, 50, true)
      printHeadTail(upstrmRsp, 35, false)
      
      --upstrmRsp = "HTTP/1.1 200 OK\r\nContent-Type: text/plain\n\nOK"

      clientSckt:send(upstrmRsp.."\r\n\r\n")
    elseif err=='timeout' then
      --print("timeout")
    else
      error("[clientSckt:error] "..tostring(err))
    end
    clientSckt:close()
  elseif err=='timeout' then
    --print("timeout")
  else
    error("[svcSckt:error] "..tostring(err))
  end
end
